### DISCUSSION TITLE
VSCode Integration - How To

### DISCUSSION THREAD

**Discussion by rsoika** [eclipse-glsp/glsp#872] [Q&A] [answered]
Currently I try to integrate my [Open-BPMN project](https://github.com/imixs/open-bpmn) - which is currently based on Theia - into VSCode. But so far I did not have success. 

By exploring the [glsp-vscode-integration project](https://github.com/eclipse-glsp/glsp-vscode-integration) I got the impression that it is recommanded to separte the VSCode-Integration into a isolated project. In the example/workflow there are two modules the *extension* and the *webview*. 

I did not understand the concept of the webview. It seems to be similar to the app module I already have in my main project which currently has the following structure

```
open-bpmn.glsp-client/
├── open-bpmn-app
├── open-bpmn-glsp
├── open-bpmn-model
├── open-bpmn-properties
└── open-bpmn-theia
```

So my first question to better understand the concept is: Is it correct/recommanded to separate the code with the vscode-webview in a extra project isolated from my main development? What if I would also separate the theia integration into a extra project to get something like this:

```
open-bpmn.glsp-client/
├── open-bpmn-glsp
├── open-bpmn-model
└── open-bpmn-properties

open-bpmn.glsp-theia/
├── open-bpmn-app
└── open-bpmn-theia

open-bpmn.glsp-vscode/
├── open-bpmn-webview
└── open-bpmn-vscode
```

I came to this question because currently I can't build the vscode-webview because of a missing  [peer-dependency](https://github.com/imixs/open-bpmn/blob/f4d2336bd4ca6c88c05b3cd4ee5f2e3ea56f6e8c/open-bpmn.glsp-client/open-bpmn-properties/package.json#L14) regarding `react-core`. This `peerDependencies` was necessary to get the JsonForms working. But the vscode-webview can't resolve this dependency for some reason.

**Comment by tortmayr** [comment, marked as answer]
>I did not understand the concept of the webview. It seems to be similar to the app module I already have in my main project which currently has the following structure

Theia extensions can freely contribute to both Theia's frontend and backend process so the entire glue code can be aggregated into one package (`open-bpmn-theia` in your case).

For VS Code extensions the story is a little bit different. VS Code extensions are running as an isolated process in the VS Code backend and don't have direct access to the fronted.
In order to contribute a Diagram Widget/View we have to use the [VS Code Webview API](https://code.visualstudio.com/api/extension-guides/webview). In a nutshell, a Webview is like an IFrame that is controlled by our extension and allows us to render arbitrary HTML content. 
So what we have to do  is bundling the code that should run in the webview and then serve it with the vscode extension.
There is no strict to do this in a separate package. You could also combine the webview and the extension package but in our experience that can get quite messy.

In the end, the webview package generates one bundled javascript file (`extension/pack/webview.js`) which is then loaded by the extension (https://github.com/eclipse-glsp/glsp-vscode-integration/blob/d437878f9002fde3800653a784d2356f5912dc38/example/workflow/extension/src/workflow-editor-provider.ts#L38)



>So my first question to better understand the concept is: Is it correct/recommanded to separate the code with the vscode-webview in a extra project isolated from my main development? What if I would also separate the theia integration into a extra project to get something like this:

We cannot make a clear recommendation for either of the two approaches. Both a perfectly valid and have their pros and cons.
**Approach 1: One big mono repo e.g.**
```
open-bpmn.glsp-client/
├── theia
        ├──open-bpmn-theia
        └── open-bpmn-app
├──vscode
        ├──open-bpmn-webview
        └── open-bpmn-vscode
├── open-bpmn-app
├── open-bpmn-glsp
├── open-bpmn-model
├──open-bpmn-properties
├── package.json (root)
└──  lerna.json
```
Has the advantage that all components are interlinked e.g. you can just use the local version of `open-bpmn-glsp` in both the Theia and VS Code integration. On the other hand you end up with a larger mono repos and version engagement and dependency resolving can become (even more) cumbersome.  Also you would probably need to adapt your build scripts so that you can also only rebuild parts of your application (i.e. the Theia integration without the vscode integration or vice versa)

**Approach 2 (Splitting into dedicated sub projects):**
```
open-bpmn.glsp-client/
├── open-bpmn-glsp
├── open-bpmn-model
└── open-bpmn-properties
└── lerna.json

open-bpmn.glsp-theia/
├── open-bpmn-app
└── open-bpmn-theia
└── lerna.json

open-bpmn.glsp-vscode/
├── open-bpmn-webview
└── open-bpmn-vscode
└── lerna.json
```
This approach is on paper the cleaner solution. All projects are clearly structured and decoupled. Unfortunately with this approach you have to take some manual action to properly link the different packages. For instance per default the Theia project is completely separated from the GLSP-client and does not know about its packages. You would need a custom process to share the packages e.g. using [`yarn link`](https://classic.yarnpkg.com/lang/en/docs/cli/link/) or a local package repository.

I personally would probably stick with approach 1.


> I came to this question because currently I can't build the vscode-webview because of a missing [peer-dependency](https://github.com/imixs/open-bpmn/blob/f4d2336bd4ca6c88c05b3cd4ee5f2e3ea56f6e8c/open-bpmn.glsp-client/open-bpmn-properties/package.json#L14) regarding `react-core`. This `peerDependencies` was necessary to get the JsonForms working. But the vscode-webview can't resolve this dependency for some reason.

You need an actual dependency to `react-core` in your webview package. With using peer dependencies you basically shift the responsibility. You declare that the `open-bpmn-properties` package requires `react-core` with a matching version range but it's up to the consuming package (the webview in this case) to actually provide the matching dependency.

**Reply by rsoika** [reply]
@tortmayr Thanks a lot! This helps me to go further. I understand the problem now better and will first resolve my problem with the `react-core` dependency and than I will try to restructure my project step be step. As Theia will be my main platform, maybe I just separate the vscode integration into a extra project to have my main codebase clean.

**Comment by rsoika** [comment]
Hi @tortmayr, I now made some progress. After cleaning up the dependencies in my main project, things look now much better. Compiling my vscode extension is successful after I added the following additional dependencies:

```
  "dependencies": {
    "react": "^18.2.0",
    "react-dom": "^18.2.0"
  },   
```

These were necessary because I use  JsonForms with the vanilla renderer and this required at lease react ^18.2.0

But now I have still some trouble when I launch the Extension in VS-Code. 

My Server jar is resolved correctly (I think) but I get the following console output in VS-Code:

```
GLSP-Server: WARNING: sun.reflect.Reflection.getCallerClass is not supported. This will impact performance.
/usr/share/code/resources/app/out/vs/workbench/api/node/extensionHostProcess.js:98
GLSP-Server: 
/usr/share/code/resources/app/out/vs/workbench/api/node/extensionHostProcess.js:98
GLSP-Server: [GLSP-Server]:Startup completed
/usr/share/code/resources/app/out/vs/workbench/api/node/extensionHostProcess.js:98
GLSP-Server: Feb. 01, 2023 11:00:41 NACHM. org.eclipse.lsp4j.jsonrpc.RemoteEndpoint fallbackResponseError
SCHWERWIEGEND: Internal error: java.lang.reflect.InvocationTargetException
java.lang.RuntimeException: java.lang.reflect.InvocationTargetException
	at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.lambda$null$0(GenericEndpoint.java:67)
	at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.request(GenericEndpoint.java:120)
	at org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.handleRequest(RemoteEndpoint.java:261)
	at org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.consume(RemoteEndpoint.java:190)
	at org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer.handleMessage(StreamMessageProducer.java:194)
	at org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer.listen(StreamMessageProducer.java:94)
	at org.eclipse.lsp4j.jsonrpc.json.ConcurrentMessageProcessor.run(ConcurrentMessageProcessor.java:113)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: java.lang.reflect.InvocationTargetException
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.lambda$null$0(GenericEndpoint.java:65)
	... 11 more
Caused by: java.lang.NullPointerException
	at java.base/java.util.Objects.requireNonNull(Objects.java:221)
	at java.base/java.util.Optional.<init>(Optional.java:107)
	at java.base/java.util.Optional.of(Optional.java:120)
	at org.eclipse.glsp.server.internal.session.DefaultClientSessionFactory.create(DefaultClientSessionFactory.java:51)
	at org.eclipse.glsp.server.internal.session.DefaultClientSessionManager.getOrCreateClientSession(DefaultClientSessionManager.java:65)
	at org.eclipse.glsp.server.protocol.DefaultGLSPServer.initializeClientSession(DefaultGLSPServer.java:120)
	... 16 more
/usr/share/code/resources/app/out/vs/workbench/api/node/extensionHostProcess.js:98
```


And Diagrams can not be loaded. 

![image](https://user-images.githubusercontent.com/1339124/216174458-e90557f3-d8e0-423c-9b40-4822a1366d46.png)


I guess my problem is the file

    /extension/src/workflow-extension.ts

I took this from the original workflow-example vscode integration.
My problem is that I can't understand the meaning of `workflowServer` and `customEditorProvider`.

Can you provide me some comments, so that I can better understand all the variables used here. Or is there some documentation I missed?

```javascript
   // Wrap server with quickstart component
    const workflowServer = new SocketGlspVscodeServer({
        clientId: 'glsp.workflow',
        clientName: 'workflow',
        serverPort: JSON.parse(process.env.GLSP_SERVER_PORT || DEFAULT_SERVER_PORT)
    });

    // Initialize GLSP-VSCode connector with server wrapper
    const glspVscodeConnector = new GlspVscodeConnector({
        server: workflowServer,
        logging: true
    });

    const customEditorProvider = vscode.window.registerCustomEditorProvider(
        'workflow.glspDiagram',
        new WorkflowEditorProvider(context, glspVscodeConnector),
        {
            webviewOptions: { retainContextWhenHidden: true },
            supportsMultipleEditorsPerDocument: false
        }
    );

    context.subscriptions.push(workflowServer, glspVscodeConnector, customEditorProvider);
    workflowServer.start();

    configureDefaultCommands({ extensionContext: context, connector: glspVscodeConnector, diagramPrefix: 'workflow' });

    context.subscriptions.push(
        vscode.commands.registerCommand('workflow.goToNextNode', () => {
            glspVscodeConnector.sendActionToActiveClient(NavigateAction.create('next'));
        }),
        vscode.commands.registerCommand('workflow.goToPreviousNode', () => {
            glspVscodeConnector.sendActionToActiveClient(NavigateAction.create('previous'));
        }),
        vscode.commands.registerCommand('workflow.showDocumentation', () => {
            glspVscodeConnector.sendActionToActiveClient(NavigateAction.create('documentation'));
        })
    );
```

**Reply by tortmayr** [reply]
Hi @rsoika,
First of all I recommend to familiarize yourself with the [VS Code extension API](https://code.visualstudio.com/api) (if you haven't done that already). VS Code uses a more restrictive declarative approach which allows you to hook into certain well defined contribution points. All of these contribution points are defined in extension [`package.json`](https://github.com/eclipse-glsp/glsp-vscode-integration/blob/d437878f9002fde3800653a784d2356f5912dc38/example/workflow/extension/package.json#L30). 

### Custom Editor Provider
The main contribution point for GLSP is the [Custom Editor API](https://code.visualstudio.com/api/extension-guides/custom-editors) which enables us to define a custom webview-based editor a specific GLSP diagram language.
If you have a more detailed look at the `package.json`'s `contributes` section you can see that we define a new `customEditor` that is associated with a specific viewType ("workflow.glspDiagram"). Implementationwise each custom editor is represented by  a [CustomEditorProvider](https://code.visualstudio.com/api/references/vscode-api#CustomEditorProvider) that encapsulates the knowledge for resolving,opening, saving etc. of the editor documents (i.e. GLSP diagrams).


### SocketGlspVscodeServer
The server setup is split into two steps:
1. Launching of the server process: this is the responsibility of the [`GlspServerLauncher`](https://github.com/eclipse-glsp/glsp-vscode-integration/blob/master/example/workflow/extension/src/workflow-extension.ts#L39)
2. Connecting to the server and initialzing the GLSP client: his is the responsibility of the [`SocketGlspVscodeServer`](https://github.com/eclipse-glsp/glsp-vscode-integration/blob/d437878f9002fde3800653a784d2356f5912dc38/example/workflow/extension/src/workflow-extension.ts#L50)


In case you missed it these concepts are also documented in the [README](https://github.com/eclipse-glsp/glsp-vscode-integration/tree/master/packages/vscode-integration#getting-started) of the `vscode-integration` package.

> But now I have still some trouble when I launch the Extension in VS-Code.
> 
> My Server jar is resolved correctly (I think) but I get the following console output in VS-Code:
> 
> ```
> GLSP-Server: WARNING: sun.reflect.Reflection.getCallerClass is not supported. This will impact performance.
> /usr/share/code/resources/app/out/vs/workbench/api/node/extensionHostProcess.js:98
> GLSP-Server: 
> /usr/share/code/resources/app/out/vs/workbench/api/node/extensionHostProcess.js:98
> GLSP-Server: [GLSP-Server]:Startup completed
> /usr/share/code/resources/app/out/vs/workbench/api/node/extensionHostProcess.js:98
> GLSP-Server: Feb. 01, 2023 11:00:41 NACHM. org.eclipse.lsp4j.jsonrpc.RemoteEndpoint fallbackResponseError
> SCHWERWIEGEND: Internal error: java.lang.reflect.InvocationTargetException
> java.lang.RuntimeException: java.lang.reflect.InvocationTargetException
> 	at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.lambda$null$0(GenericEndpoint.java:67)
> 	at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.request(GenericEndpoint.java:120)
> 	at org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.handleRequest(RemoteEndpoint.java:261)
> 	at org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.consume(RemoteEndpoint.java:190)
> 	at org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer.handleMessage(StreamMessageProducer.java:194)
> 	at org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer.listen(StreamMessageProducer.java:94)
> 	at org.eclipse.lsp4j.jsonrpc.json.ConcurrentMessageProcessor.run(ConcurrentMessageProcessor.java:113)
> 	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
> 	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
> 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
> 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
> 	at java.base/java.lang.Thread.run(Thread.java:829)
> Caused by: java.lang.reflect.InvocationTargetException
> 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
> 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
> 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
> 	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
> 	at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.lambda$null$0(GenericEndpoint.java:65)
> 	... 11 more
> Caused by: java.lang.NullPointerException
> 	at java.base/java.util.Objects.requireNonNull(Objects.java:221)
> 	at java.base/java.util.Optional.<init>(Optional.java:107)
> 	at java.base/java.util.Optional.of(Optional.java:120)
> 	at org.eclipse.glsp.server.internal.session.DefaultClientSessionFactory.create(DefaultClientSessionFactory.java:51)
> 	at org.eclipse.glsp.server.internal.session.DefaultClientSessionManager.getOrCreateClientSession(DefaultClientSessionManager.java:65)
> 	at org.eclipse.glsp.server.protocol.DefaultGLSPServer.initializeClientSession(DefaultGLSPServer.java:120)
> 	... 16 more
> /usr/share/code/resources/app/out/vs/workbench/api/node/extensionHostProcess.js:98
> ```
I agree, to me it looks like the server is resolved and started correctly. Unfortunatley, I have no idea what's going wrong here.
I would recommend, to debug into the failing method. The should be a [launch config](https://github.com/eclipse-glsp/glsp-vscode-integration/blob/d437878f9002fde3800653a784d2356f5912dc38/.vscode/launch.json#L23) that allows you to start the server externally (e.g. in Eclipse) for debuging purposes

**Reply by rsoika** [reply]
Ok thanks, I will investigate into this further... I have tested it with the GLSP Workflow example before, which of course worked. So I think it is something with my configuration. And I think `clientID` and `clientName` must match somehow to my Server and/or Client code. But did not found these parameters yet.

**Comment by rsoika** [comment]
Hi @tortmayr,
once again I come back to the issue 'How to integrate a GLSP Diagram into VSCode'

I tried to read carefully all the documentation. I also checked all my configuration several times. 
I know, that my server code is integrated correctly because I can debug it when I start the VSCode extension with a external server (started in Eclipse Debug mode).

I am also pretty sure that my `customEditors` in `package.json` is correct because I can see that in VSCode the open with... menu shows my Open-BPMN extension

```
    "customEditors": [
      {
        "viewType": "bpmn-diagram",
        "displayName": "BPMN 2.0 Diagram Editor",
        "selector": [
          {
            "filenamePattern": "*.bpmn"
          }
        ]
      }
    ],
```

I guess the parameter `viewType` is the one which is mapped to the  customEditorProvider registration

```javascript
    const customEditorProvider = vscode.window.registerCustomEditorProvider(
        'bpmn-diagram',
        new BPMNEditorProvider(context, glspVscodeConnector),
        {
            webviewOptions: { retainContextWhenHidden: true },
            supportsMultipleEditorsPerDocument: false
        }
    );
```

To me all looks correct so far. But I still get only a blank page in vs-code:

![image](https://user-images.githubusercontent.com/1339124/216848451-ac4d9887-27c8-461c-89db-4ba62f85837f.png)

I can see that my `BPMNlowEditorProvider` is loaded and also the `webviewPanel.webview.html`.

Now I wonder if my `BPMNlowEditorProvider` can be the problem which extends the `GlspEditorProvider` from the `quickstart-components`


```javaScript
export default class BPMNlowEditorProvider extends GlspEditorProvider {
    diagramType = 'bpmn-diagram';
 ....
}
```

(By the way, I still did not get the difference between `viewType` and `diagramType` , this is why I set them to the same values)

Is it possible that in my case (no GLSP Json Model) I have to implement the methods `openCustomDocument` and `saveCustomDocument` methods by myself?

I also published my code on github: https://github.com/imixs/open-bpmn-vscode-integration/tree/main/open-bpmn.vscode

**Comment by rsoika** [comment]
I made again some progress. 
I figured out that one critical part seems to be again the sprotty dependency. 

Currently the root package.json file looks like this:

```
{
  "private": true,
  "engines": {
    "yarn": ">=1.7.0 <2.x.x",
    "node": ">=14.18.0"
  },
  "scripts": {
    "all": "yarn install && yarn lint",
    "build": "lerna run build",
    "clean": "lerna run clean",
    "lint": "lerna run lint",
    "lint:ci": "lerna run lint:ci",
    "prepare": "lerna run prepare",
    "publish:prepare": "lerna version --ignore-scripts --yes --no-push",
    "publish:latest": "lerna publish from-git --no-git-reset --no-git-tag-version --no-verify-access --no-push",
    "publish:next": "SHA=$(git rev-parse --short HEAD) && lerna publish preminor --exact --canary --preid next.${SHA} --dist-tag next --no-git-reset --no-git-tag-version --no-push --ignore-scripts --yes --no-verify-access",
    "upgrade:next": "yarn upgrade -p \"@eclipse-glsp.*|sprotty.*\" --next ",
    "watch": "lerna run --parallel watch"
  },
  "dependencies": {
    "react": "^18.2.0",
    "react-dom": "^18.2.0"
  },
  "devDependencies": {
    "@eclipse-glsp/config": "1.1.0-next.e31f2e5.119",
    "@eclipse-glsp/config-test": "1.1.0-next.e31f2e5.119",
    "@types/node": "16",
    "@types/vscode": "^1.54.0",
    "mvn-artifact-download": "5.1.0"
  },
  "resolutions": {
    "**/sprotty": "0.13.0-next.f4445dd.342"
  },  
  "workspaces": [
    "packages/*/",
    "open-bpmn.vscode/*"
  ]
}
```

During compile time I see the following warning:

`warning Resolution field "sprotty@0.13.0-next.f4445dd.342" is incompatible with requested version "sprotty@~0.12.0"`

Can this be a possible root of my problems?

**Comment by rsoika** [comment]
@tortmayr 
You don't need to take again a look at this problem. Sorry for taking your time. 
The sprotty dependencies is irritating, but maybe not the root of my problems. It seems to be possible somewhere else in my own server code.....

I need to do another debugging session first and try to solve the problem maybe by myself. Or at least I will be able to post my problem with more useful details...

**Comment by rsoika** [comment]
Hi @tortmayr,
good news - the VS-Code integration works! (80%)
I do not know what the problem was at the end. 

 - maybe my settings for `diagramType` and `viewType`  were wrong in the beginning. I aligned these types to the `getDiagramtype()` method in the `DiagramModule`. 
 - the sprotty dependecy seems also to be important (as always). It works if I use `resolutions` in the root package.json file (even if I got a compiler warning):

```
  "resolutions": {
    "**/sprotty": "0.13.0-next.f4445dd.342"
  },  
```

And finally I needed to change the implementation of my `SourceModelStorage` (which I am using because I am not working on a JSON-Model).

I used the option `uri` to get the filename to be opened. But VSCode Integration only provides `sourceUri` which seems to be the correct one in both cases - for GLSP and for VSCode.

**Note** the case sensitive name of  `sourceUri`. I think the documentation [here](https://www.eclipse.org/glsp/documentation/sourcemodel/) is not correct. It is using the option name `sourceURI`. But I think it should be always `sourceUri` ...?
